### PR TITLE
google: adding Google DNS module

### PR DESCRIPTION
**Pull request by erjohnso**
Since this is a new module, starting it out in 'extras'.  This module is based heavily on the route53 module in order to make it easier for ansible users to move between the two cloud providers.

_WARNING_: This pull-request depends on https://github.com/ansible/ansible/pull/9452.

/cc @mpdehaan - per our agreement, I'm asking for a review and initial merge from you (or someone else you designate).  Thanks!

### CODE REVIEW

**Comment by erjohnso** [comment]
Per discussion with Michael, I'll be making several changes to this.  So, even though https://github.com/ansible/ansible/pull/9452 is now merged, please do not merge this yet.

**Comment by bcoca** [comment]
could you move this into a 'google' subdirectory (as we have in core)?

**Comment by paparomeo** [comment]
Ping! I am really interested in this module being merged, @erjohnso. Can I help in any capacity?

File: cloud/gc_dns.py [hunk 1]
```diff
@@ -0,0 +1,349 @@
+#!/usr/bin/env python
```
Comment (bcoca) [inline cloud/gc_dns.py:1]: shebang should be /usr/bin/python

File: cloud/gc_dns.py [hunk 2]
```diff
@@ -0,0 +1,349 @@
...
+DOCUMENTATION = '''
+---
+module: gc_dns
+version_added: "1.8"
+short_description: Manage Google Cloud DNS resources
+description:
+     - Manage Google Could DNS resources.
+       U(https://cloud.google.com/products/cloud-dns) for an overview.
+options:
+  command:
```
Comment (bcoca) [inline cloud/gc_dns.py:28]: we prefer state=present/absent, leaving listing/info for facts modules

File: cloud/gc_dns.py [hunk 3]
```diff
@@ -0,0 +1,349 @@
...
+            # OK that this is already absent, supports idempotency
+            output['name'] = name
+        except Exception, e:
+            if e.http_code == 404:
+                pass
+            else:
+                module.fail_json(msg=unexpected_error_msg(e), changed=False)
+
+    else:
+        module.fail_json(msg="Unknown command '%s', must be 'list', 'get', 'create', or 'delete'." % command,
```
Comment (bcoca) [inline cloud/gc_dns.py:336]: not needed as module definition should already detect this

File: cloud/gc_dns.py [hunk 4]
```diff
@@ -0,0 +1,349 @@
...
+    from libcloud.dns.providers import get_driver
+    from libcloud.common.google import (GoogleBaseAuthConnection,
+                                        ResourceExistsError,
+                                        GoogleInstalledAppAuthConnection,
+                                        GoogleBaseConnection)
+    _ = Provider.GOOGLE
+except ImportError:
+    print("failed=True " + \
+        "msg='libcloud with GCE support >=0.15.0 required for this module'")
+    sys.exit(1)
```
Comment (bcoca) [inline cloud/gc_dns.py:159]: use fail_json from module instead, we normally defer the error until module is instantiated to use this facility

File: cloud/gc_dns.py [hunk 5]
```diff
@@ -0,0 +1,349 @@
...
+
+    else:
+        module.fail_json(msg="Unknown command '%s', must be 'list', 'get', 'create', or 'delete'." % command,
+                changed=False)
+
+    output['command'] = command
+    if 'changed' not in output:
+        output['changed'] = False
+    print json.dumps(output)
+    sys.exit(0)
```
Comment (bcoca) [inline cloud/gc_dns.py:343]: again, use the module object's facilities, exit_json should take care of this.

**Comment by gregdek** [comment]
@erjohnso if you can fix the issues brought up by @bcoca we're happy to merge this.

File: cloud/gc_dns.py (hunk 2 above)
Comment (tima) [inline cloud/gc_dns.py:28]: +1

**Comment by erjohnso** [comment]
Yes, sorry I've let this rot so long!  I'll shoot for the end of this week.

File: cloud/gc_dns.py (hunk 1 above)
Comment (ipv1337) [inline cloud/gc_dns.py:1]: Why use the hardcoded /usr/bin/python over the more flexible alternative what like Eric used? The path /usr/bin/env will always exist except in really rare cases.

Comment (bcoca) [inline cloud/gc_dns.py:1]: http://docs.ansible.com/ansible/faq.html#how-do-i-handle-python-pathing-not-having-a-python-2-x-in-usr-bin-python-on-a-remote-machine

**Comment by gregdek** [comment]
Adding new process. We will be evaluating all new module PRs according to this process, effective immediately.

# 

Thanks for submitting this new module to Ansible Extras! This module is now in community review, a process that is open to all Ansible users. In order for this module to be approved, it must gain the following votes:

“works_for_me”: If you have tested the module thoroughly, including testing of all of the module’s options, and if the module works for you, please add “works_for_me” in the comments.

“passes_guidelines”: If you have gone through the module guidelines and the module meets all of the requirements, please add “passes_guidelines” in the comments. Guidelines are available here: http://docs.ansible.com/developing_modules.html#module-checklist

“needs_revision”: If the module fails to work for you, or if it doesn’t meet guidelines, please add “needs_revision” in the comments with details about what needs to be fixed.

When a module has both “works_for_me” and “passes_guidelines” tags, we will promote the module for inclusion in Ansible Extras.  At this point, you will be expected to maintain the module by fixing bugs and evaluating pull requests in a timely manner. 

Thanks again for submitting your Ansible module!

**Comment by gregdek** [comment]
@erjohnso A friendly reminder: this pull request has been marked as needing your action, and it hasn't been updated in the last month. 

If you still believe that this PR applies, and you intend to address the issues with this PR, just let us know in the PR itself and we'll keep it open pending your changes.

If we receive no indication from you within a week, we will close this pull request.

**Comment by gregdek** [comment]
Closing due to contributor inactivity. 

@erjohnso if/when you're ready to make the needed revisions, feel free to reopen and make your changes, and we'll get it back into the review process.

**Comment by michaelbannister** [comment]
@erjohnso I was trying to do something along the same lines when I found this PR; the change you made in ansible/ansible#9452 doesn't seem to work, as the gce_connect function calls the get_driver function from libcloud.compute.providers which fails because the provider 'google' is not known.

**Comment by walbert947** [comment]
@michaelbannister We ran into this problem with our own independent in-house Google DNS module. The gce_connect module was hard-coded to load Provider and get_driver from the libcloud.compute namespace, which breaks the DNS module because it expects Provider and get_driver from the libcloud.dns namespace.

My workaround was to call the appropriate libcloud imports again _after_ importing ansible.module_utils.gce. This will override the hard-coded libcloud.compute imports with libcloud.dns.

Like so:

``` python
# <...snip...>

from ansible.module_utils.gce import *
# Insert check to see if libcloud is actually available here
from libcloud.dns.types import Provider
from libcloud.dns.providers import get_driver

if __name__ == '__main__':
    main()
```

A dirty hack, but I guess I'm partly to blame for sitting on a private module for so long without contributing it upstream. :stuck_out_tongue:

**Comment by erjohnso** [comment]
@michaelbannister, you're also (in addition to @walbert947) using this DNS module even though it didn't get merged upstream?

Going way back in time, I had tried to model it after the route53 module. But that wasn't idiomatic Ansible, and the feedback from Michael and Brian was pretty clear this version wasn't going to make it.  And of course, I really dropped the ball on not getting this fixed up (or closed with replacement).

**Comment by walbert947** [comment]
@erjohnso I built an original module about a year ago, because our team was using Ansible to create instances in GCE and was getting bogged down by having to create DNS entries for them by hand. I didn't see an existing module in Ansible, and didn't think to check the list of open PRs before writing my own. Oops, heh.

Comparing our code, my module (which is split up between records and zones) is more idiomatic Ansible, and it also supports check mode. I also have a comprehensive testing playbook for it. However, I made no attempt at preserving compatibility with the Route53 module.

I've been meaning to contribute my module upstream, but I wanted to let it bake internally for a bit, and just got sidetracked by other projects. Since it looks like there's still some community demand for a Google DNS module, I may move forward with a new PR if this one isn't going anywhere in the near future.

**Comment by erjohnso** [comment]
@walbert947, I had originally assumed parity with the route53 module would be desirable, but that seems to not be the case (and the route53 module may have diverged from its original design too for all I know).

This PR isn't going to get picked up again, so I think starting fresh with a more idiomatic module is the right thing to do. I would be very grateful if you sent in the PR for your module and would be happy to review / test.

I would actually be so grateful if you contributed the module, I'd be happy to mail you a Google Cloud Platform t-shirt if that would be of interest. :)

**Comment by walbert947** [comment]
@erjohnso Sure, that'd be great. :)

Let me go do a bit of cleanup, and then I'll open up a new PR (hopefully by the weekend).

**Comment by erjohnso** [comment]
@walbert947, awesome and thank you! Feel free to /cc me on the PR.

Shoot me an email with size and address erjohnso -at- google -dot- com.

**Comment by michaelbannister** [comment]
This sounds like good news! :)

I am relatively new to Ansible and have been looking at automating setup of a site on GCE as a way of learning it. I made a start on my own module and got as far as a "zone" task (I also don't really have any experience with Python) before getting distracted onto something else…

@walbert947 I'll be really interested to see your testing playbook as all I could really find in the "Developing Modules" documentation was a few brief mentions of ansible/hacking/test-module, which seemed a bit ad-hoc.

**Comment by walbert947** [comment]
@michaelbannister I was primarily focused on integration tests, as the entire purpose of the Google Cloud DNS module was to serve as an Ansible-y wrapper around Google DNS's API. There are good examples in ansible/test/integration/roles.

**Comment by walbert947** [comment]
Sorry for the delay @erjohnso. The module was originally developed for Ansible 1.9, so there was a bit more work involved than I expected in porting it.

The new Google DNS module PR is at https://github.com/ansible/ansible-modules-extras/pull/2252.

**Comment by erjohnso** [comment]
Awesome, thank you very much @walbert947! I'll give it a review.

**Pull request closed without merging**